burn-core/src/nn/loss/cross_entropy.rs

```diff
@@ -25,85 +97,451 @@ impl<B: Backend> CrossEntropyLoss<B> {
+                let tensor = tensor.reshape([batch_size]) * weights.clone();
+                let tensor = Self::apply_mask_1d(tensor, mask);
+                tensor.sum().neg() / weights.sum()
+            }
+            None => {
+                let tensor = Self::apply_mask_1d(tensor.reshape([batch_size]), mask);
+                tensor.mean().neg()
+            }
+        }
+    }
 
-        tensor.mean().neg()
+    fn compute_smoothed_targets(
+        shape: [usize; 2],
+        targets: Tensor<B, 1, Int>,
+        alpha: f32,
+    ) -> Tensor<B, 2> {
+        let [batch_size, nr_classes] = shape;
+        let device = &targets.device();
+        let targets_matrix = Tensor::<B, 2>::zeros_device(shape, device).scatter(
+            1,
+            targets.reshape([batch_size, 1]),
+            Tensor::ones_device([batch_size, 1], device),
+        );
+        targets_matrix * (1. - alpha) + alpha / nr_classes as f32
     }
 
     fn padding_mask(&self, targets: &Tensor<B, 1, Int>) -> Option<Tensor<B, 1, Bool>> {
         let mut mask = None;
-        if let Some(pad_index) = self.pad_index {
-            mask = Some(targets.clone().equal_elem(pad_index as i64));
+        if let Some(pad_tokens) = &self.pad_tokens {
+            let mut res = targets.clone().equal_elem(pad_tokens[0] as i64).int();
+            for x in pad_tokens {
+                res = res + targets.clone().equal_elem(*x as i64).int();
+            }
+            mask = Some(res.greater_elem(0));
         }
 
         mask
     }
 
-    fn apply_mask(
-        &self,
-        mut tensor: Tensor<B, 1>,
-        mask: Option<Tensor<B, 1, Bool>>,
-    ) -> Tensor<B, 1> {
+    fn apply_mask_1d(mut tensor: Tensor<B, 1>, mask: Option<Tensor<B, 1, Bool>>) -> Tensor<B, 1> {
         if let Some(mask) = mask {
             tensor = tensor.mask_fill(mask, 0);
         }
 
         tensor
     }
+
+    fn apply_mask_2d(mut tensor: Tensor<B, 2>, mask: Option<Tensor<B, 1, Bool>>) -> Tensor<B, 2> {
+        if let Some(mask) = mask {
+            let [batch_size, nr_classes] = tensor.dims();
+            tensor = tensor.mask_fill(mask.reshape([batch_size, 1]).repeat(1, nr_classes), 0);
+        }
+
+        tensor
+    }
+
+    fn assertions(logits: Tensor<B, 2>, targets: Tensor<B, 1, Int>) {
+        let [logits_height, logits_width] = logits.dims();
+        let [targets_height] = targets.dims();
+        assert!(
+            logits_height == targets_height,
+            "Shape of targets ({}) should correspond to outer shape of logits ({}).",
+            targets_height,
+            logits_height
+        );
+        let max_target = targets.clone().max();
+        assert!(
+            max_target
+                .clone()
+                .lower_elem(logits.dims()[1] as i32)
+                .into_data()
+                .value[0],
+            "Encounter target ({}) greater than feature dimension ({}).",
+            max_target,
+            logits_width
+        );
```

Comment:
I would remove that assertion, asserting shapes is fine, but when it comes to the underlying tensor data, it's normally too slow.